common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java

```diff
@@ -368,4 +381,70 @@ private static void closeQuietly(Closeable c) {
+        static final Method HAS_OPTION_METHOD;
+
+        static {
+            Method toPathMethod = null;
+            Method getFileStoreMethod = null;
+            Method mountEntryMethod = null;
+            Method hasOptionMethod = null;
+            try {
+                // Use reflection so Java 6 fails gracefully.
+                toPathMethod = File.class.getMethod("toPath");
+                Class<?> pathClass = Class.forName("java.nio.file.Path");
+                getFileStoreMethod = Class.forName("java.nio.file.Files").getMethod("getFileStore", pathClass);
+
+                mountEntryMethod = Class.forName("sun.nio.fs.UnixFileStore").getDeclaredMethod("entry");
+                hasOptionMethod =
+                        Class.forName("sun.nio.fs.UnixMountEntry").getDeclaredMethod("hasOption", String.class);
+
+                Throwable mountEntryThrowable = ReflectionUtil.trySetAccessible(mountEntryMethod);
+                Throwable hasOptionThrowable = ReflectionUtil.trySetAccessible(hasOptionMethod);
+                if (mountEntryThrowable != null || hasOptionThrowable != null) {
+                    if (!(mountEntryThrowable instanceof SecurityException) ||
+                        !(hasOptionThrowable instanceof SecurityException)) {
+                        logger.debug("Failed to initialize NativeLibraryLoader.NoexecFileStoreDetector; " +
+                                     "use --add-opens java.base/sun.nio.fs=... to get noexec mount warnings");
+                    }
+                    mountEntryMethod = null;
+                    hasOptionMethod = null;
+                    // Re-throw the security exception to log it
+                    throw mountEntryMethod == null ? hasOptionThrowable : mountEntryThrowable;
+                }
+            } catch (ClassNotFoundException e) {
+                // Normal on non-Unix operating systems or on Java < 7
+            } catch (Throwable t) {
+                logger.debug("Failed to initialize NativeLibraryLoader.NoexecFileStoreDetector", t);
+            }
+            TO_PATH_METHOD = toPathMethod;
+            GET_FILE_STORE_METHOD = getFileStoreMethod;
+            MOUNT_ENTRY_METHOD = mountEntryMethod;
+            HAS_OPTION_METHOD = hasOptionMethod;
+        }
+
+        static boolean isOnNoexecFileStore(File file) {
+            if (HAS_OPTION_METHOD == null || MOUNT_ENTRY_METHOD == null ||
+                GET_FILE_STORE_METHOD == null || TO_PATH_METHOD == null) {
+                return false;
+            }
+            try {
+                Object path = TO_PATH_METHOD.invoke(file);
+                Object fileStore = GET_FILE_STORE_METHOD.invoke(null, path);
+                Object mountEntry = MOUNT_ENTRY_METHOD.invoke(fileStore);
+                return (Boolean) HAS_OPTION_METHOD.invoke(mountEntry, "noexec");
+            } catch (Throwable t) {
+                logger.debug("Error checking if {} is on a file store mounted with noexec", file, t);
```

Comment:
can we let this method throw and catch/log at a higher level? Currently we just catch a `Throwable` above and ignore it ... seems more intuitive to handle this in a single spot and if higher layers want to do something with the exception they can.